File: 10-12-2024_Find_Longest_Special_Substring_That_Occurs_Thrice_I.py
class Solution:
    def maximumLength(self, s: str) -> int:
        freq_map = {}
        n = len(s)

        # Generate all special substrings and count their frequencies
        for i in range(n):
            cur_string = ""
            for j in range(i, n):
                # Check if the substring can continue to be "special"
                if not cur_string or s[j] == cur_string[-1]:
                    cur_string += s[j]
                    freq_map[cur_string] = freq_map.get(cur_string, 0) + 1
                else:
                    break

        # Find the maximum length of a substring occurring at least 3 times
        max_len = -1
        for substring, freq in freq_map.items():
            if freq > 2:
                max_len = max(max_len, len(substring))

        return max_len

File: 10-12-2024_Find_Longest_Special_Substring_That_Occurs_Thrice_I.cpp
class Solution {
public:
    int maximumLength(std::string s) {
        std::unordered_map<std::string, int> map;
        int n = s.length();

        for (int i = 0; i < n; i++) {
            std::string curString = "";
            for (int j = i; j < n; j++) {
                // Check if the substring can continue to be "special"
                if (curString.empty() || s[j] == curString.back()) {
                    curString += s[j];  // Append the current character
                    map[curString]++;
                } else {
                    break;  // Break if the character is different
                }
            }
        }

        int maxLen = -1;
        for (const auto& entry : map) {
            const std::string& str = entry.first;
            int freq = entry.second;
            if (freq > 2) {
                maxLen = std::max(maxLen, (int)str.length());
            }
        }

        return maxLen;
    }
};
